### DISCUSSION TITLE
Action will work in development but not in production

### DISCUSSION THREAD

**Discussion by thanosoncode** [vercel/next.js#50370] [App Router]
Any ideas why this deletePost action works in development but not in production?
In production I'm getting a ReferenceError: id is not defined

```
const Home = async () => {
  const posts = (await prisma.post.findMany()) as Post[];

  return (
    <div>
      <h1 className="text-2xl mb-8">All posts</h1>
      <section className="flex flex-col gap-3">
        {posts.map((post) => (
          <div key={post.id} className="flex gap-4 justify-between items-start">
            <Link href={`/posts/${post.id}`} className="block underline">
              {post.title}
            </Link>
            {/* @ts-expect-error Server Component */}
            <DeleteForm id={post.id} />
          </div>
        ))}
      </section>
    </div>
  );
};
```

```
const DeleteForm = async ({ id }: { id: string }) => {
  const deletePost = async () => {
    "use server";
    await prisma.post.delete({ where: { id } });
    revalidatePath("/");
  };
  return (
    <form action={deletePost}>
      <button className="text-sm" type="submit">
        delete
      </button>
    </form>
  );
};
```

**Comment by icyJoseph** [comment]
What Next.js version have you got? It doesn't work for me, in either dev nor production mode.

I imagine the issue has to do with closure capturing, and how that would have been a way to leak secrets (it was actually shown to do that). [YouTube video](https://www.youtube.com/watch?v=fdKcbyEK66M)

For now I guess you gotta refactor to using a client component in the deletion bit:

```tsx
"use client";

import { deletePost } from "./actions";

export const DeleteForm = ({ id }: { id: string }) => {
  return (
    <form>
      <button
        className="text-sm"
        type="submit"
        formAction={() => deletePost({ id })}
      >
        delete
      </button>
    </form>
  );
};
```

Or! have a hidden unchanging field in that form, which contains the id to delete...

```tsx
import { deletePost } from "./actions";

export const DeleteForm = ({ id }: { id: string }) => {
  async function deleteAction(data: FormData) {
    "use server";

    const postId = data.get("postId");
    if (!postId) return;

    await deletePost({ id: postId.toString() });
  }

  return (
    <form action={deleteAction}>
      <button className="text-sm" type="submit">
        delete
      </button>
      {/* users can still change this value through the DevTools */}
      {/* it might be a good idea to hash it and un-hash it server side*/}
      <input name="postId" type="hidden" defaultValue={id} readOnly />
    </form>
  );
};
```

> I noticed a funny bug while doing this... If the hidden field is the first child of the form, then you get two input fields!

Anyway, this has the downside that people could mess with the id on devTools and what not. I think in general you kind of don't want to leak the id to the client. You could hash the id, so that it makes it harder to "guess" other ids and potentially mess things up. Maybe I am just being too careful though...

Here's an example:

- encrypt/decrypt functions https://github.com/icyJoseph/next-battlesnake/blob/main/src/utils.ts
- here I encrypt every game uuid, and send that to the client instead
- here, when a user wants to see more of a game, I take the encrypted uuid, decrypt it, and see if that exists on the db https://github.com/icyJoseph/next-battlesnake/blob/main/src/pages/api/stats/%5Buuid%5D.ts

If someone just tries to guess game id's they'll have a hard time actually generating one that, when decrypted is something that exists on my db.
